crates/ra_hir_ty/src/_match.rs

```diff
@@ -530,6 +532,16 @@ pub(crate) fn is_useful(
     matrix: &Matrix,
     v: &PatStack,
 ) -> MatchCheckResult<Usefulness> {
+    // Handle the special case of enums with no variants. In that case, no match
+    // arm is useful.
+    if let Some(enum_variant) = cx.infer.variant_resolution_for_expr(cx.match_expr) {
+        if let VariantId::EnumVariantId(enum_variant_id) = enum_variant {
+            if cx.db.enum_data(enum_variant_id.parent).variants.len() == 0 {
```

Comment:
this can never be true, since you got the enum ID from one of its variants ;)

The problem here is that `variant_resolution_for_expr` is not the thing you need. It records the specific variant apreferred to by e.g. a path or record literal. If you have something like `Ok()`, the type of that is just `Result`, so we need to record the information that it was the `Ok` variant somewhere, and that's this field. But here you don't want the variant, you want the type, which you get through `cx.infer[cx.match_expr]`. To get the enum def ID, you could then use `ty.as_adt()` and match the `AdtId` (though we might need to unwrap references -- still not sure what the best way to deal with that is, maybe we should just generally strip all references from types in match checking...)

Reply:
Ah, this makes sense. Thanks for setting me on the right track!